### PR TITLE
Cosmos OSS MI

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description

- Adds a `UseManagedIdentity` option to config
- Adds role definition assignment to ARM template
- Gains
  - Application can use MI to connect to CosmosDB to read and write on the data plane
- Impacts:
  - Throughput settings can no longer be modified
  - Stored procedures can no longer be installed
  - Partitions can no longer be read 

Note: The Integration tests will setup and tear down test collections which requires more privileges than what MI can currently provide, for this reason the Host and Key are will written to the KeyVault in the ARM template.

## Related issues
Addresses AB#96632, #3249

## Testing
E2E tests should all still run to verify connection and existing functionality.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs [hunk 1]
```diff
@@ -134,7 +152,8 @@ public async Task InitializeDataStoreAsync(CosmosClient client, CosmosDataStoreC
                                 cancellationToken: cancellationToken));
                 }
 
-                foreach (var collectionInitializer in collectionInitializers)
+                // Manged identity does not have permission to read/update stored procedures or change collection settings
+                foreach (ICollectionInitializer collectionInitializer in collectionInitializers)
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs:156]: Should this section have an if statement checking for using managed identity?

**Comment by LTA-Thinking** [comment]
Should we have a plan/method in place for handling schema changes in Cosmos before we merge this change in? I assume for OSS the steps would be to turn MI off, apply change, then turn it back on.

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs (hunk 1 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs:156]: There are some initializers that populate data

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/StoredProcedures/StoredProcedureInstaller.cs [hunk 2]
```diff
@@ -9,23 +9,33 @@
...
+            _configuration = configuration;
             _storedProcedures = storedProcedures;
         }
 
         public async Task ExecuteAsync(Container container, CancellationToken cancellationToken)
         {
+            if (_configuration.UseManagedIdentity)
+            {
+                // Managed Identity does not support read/write stored procedures
+                return;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/StoredProcedures/StoredProcedureInstaller.cs:36]: Should log then this skips

**Comment by brendankowitz** [comment]
Closing until we revisit and look into installing SPs and other dependencies

**Pull request closed without merging**